### ISSUE TITLE
Digest Auth issue with WinHttpHandler in .net core 2.2 deployed on Azure App Service

### ISSUE THREAD

**Issue by arkoc** [dotnet/corefx#37607]
This code is trying to call endpoint protected by Digest Authentication in asp.net core 2.2:
```c#
var credCache = new CredentialCache
{
    { baseUrl, "Digest", new NetworkCredential(username, password) },
};

var httpMessageHandler = new WinHttpHandler() { ServerCredentials = credCache };
var httpClient = new HttpClient(httpMessageHandler)
{
    BaseAddress = baseUrl,
};

var info = await httpClient.GetAsync("getinfo");
```

This works fine on local machine (Win 10.0.17763). But when we deploy it to Azure AppService (running .net core 2.2 image) it is returning 401.

**Note:** We use `WinHttpHandler` because new `SocketHttpHandler` was returning 401 in local machine too. 

Here is captured traffic:

**Local Machine**
![local](https://user-images.githubusercontent.com/11096006/57578468-2ec79f80-749e-11e9-9d87-b6c052fdbe5b.png)

**App Service**
![appservice](https://user-images.githubusercontent.com/11096006/57578469-3129f980-749e-11e9-8cb6-7dab03a0a2f0.png)

Those are request/responses for working workflow in local machine:

**Request I**
> GET xxxx/getinfo HTTP/1.1
> Connection: Keep-Alive
> Host: xxxx

**Response I**
> HTTP/1.1 401 Unauthorized
> Server: Epee-based
> Content-Length: 98
> Content-Type: text/html
> Last-Modified: Sun, 12 May 2019 06:17:40 GMT
> Accept-Ranges: bytes
> WWW-authenticate: Digest qop="auth",algorithm=MD5,realm="xxxx",nonce="O6oE/K7WVZzu5wk5eOqPNw==",stale=false
> WWW-authenticate: Digest qop="auth",algorithm=MD5-sess,realm="xxxx",nonce="O6oE/K7WVZzu5wk5eOqPNw==",stale=false

**Request II**
>GET xxxx/getinfo HTTP/1.1
> Connection: Keep-Alive
> Host: xxxx
> Authorization: Digest username="admin",realm="monero-rpc",nonce="O6oE/K7WVZzu5wk5eOqPNw==",uri="/getinfo",cnonce="9834e1cbc4cdea5eca55c9472a0c406b",nc=00000001,algorithm=MD5,response="b5af17dffad40bb7c71855f24bf2c39a",qop="auth"

**Response II**
> HTTP/1.1 200 Ok

**Environment Info**
.net core 2.2
WinHttpHandler nuget version: 4.5.3

**Comment by PurpleTheBest** [comment]
Same issue. @arkoc are you implementing Monero auth too?

**Comment by wfurt** [comment]
Digest Auth uses md5 and that is in process of deprecation and removal from OS for long time. 
Are you sure MD5 is enabled on your azure service? 

Can you share trace when it fails with default handler?

**Comment by arkoc** [comment]
This is the code with default handler, running it in local machine.
```c#
var credCache = new CredentialCache
{
    { baseUrl, "digest", new NetworkCredential(username, password) },
};
var httpClient = new HttpClient(new HttpClientHandler()
{
    Credentials = credCache,
});
httpClient.BaseAddress = baseUrl;
var info = await httpClient.GetAsync("getinfo");
```
This is the network trace:
![image](https://user-images.githubusercontent.com/11096006/57740023-e2ab7380-76c6-11e9-98f3-31deb5041fb2.png)

Interesting thing is when I open fiddler (nothing special, just default configs) this thing starts working, here is network trace of it:

![new](https://user-images.githubusercontent.com/11096006/57740239-d542b900-76c7-11e9-937d-584825f27e4d.png)


> Are you sure MD5 is enabled on your azure service?
I didn't get this part. Is there any specific configs to do for it?

**Comment by wfurt** [comment]
Thanks. Initial suspicion would be multiple WWW-authenticate:Digest headers. 

Could you also do packer capture instead of (or also with) Fiddler? I'm wondering if Fiddler can reorder headers or alter headers for what ever reason.

**Comment by babgev** [comment]
We stumbled to this strange behavior while testing this with the default handler 
![asyncIssue](https://user-images.githubusercontent.com/6187131/57741773-5b163280-76cf-11e9-8c0b-a025717655c8.gif)

**Comment by arkoc** [comment]
So second picture is with packet capture with Fiddler: which worked. The first one is without fiddler. Did I get your question correct, or by "packer capture" you meant anything else?

**Comment by babgev** [comment]
We stumbled into this strange behavior while debugging it with the default handler.
I enabled .NET source stepping, then stepped the calls, and as you can see in the GIF I'm moving through calls with some intentional 'waits'.

![asyncIssue](https://user-images.githubusercontent.com/6187131/57742090-c3b1df00-76d0-11e9-9fc3-4201d58a4d86.gif)

**Comment by wfurt** [comment]
Get traces with https://www.wireshark.org/#download
If you want to use loopback, you will need the ncap driver (as Windows do not show loopback as interface by default)

**Comment by cege7480** [comment]
> Get traces with https://www.wireshark.org/#download
> If you want to use loopback, you will need the ncap driver (as Windows do not show loopback as interface by default)

Scroll up.

**Comment by arkoc** [comment]
Maybe this can shad some light into this issue. If you look at succeeded request by fiddler:

![image](https://user-images.githubusercontent.com/11096006/57754422-7436d780-76ff-11e9-8112-9d450b6fd04f.png)

Second request contains exact same nonce, but without Fiddler it is different:

![image](https://user-images.githubusercontent.com/11096006/57754488-9d576800-76ff-11e9-9221-8f9e23121680.png)

**Comment by davidsh** [comment]
>This code is trying to call endpoint protected by Digest Authentication in asp.net core 2.2:

Can you post the source code for the "server" you are using that is requiring Digest?  Is it using Kestrel or IIS hosting for example?

**Comment by arkoc** [comment]
> > This code is trying to call endpoint protected by Digest Authentication in asp.net core 2.2:
> 
> Can you post the source code for the "server" you are using that is requiring Digest? Is it using Kestrel or IIS hosting for example?

It's not .net core. Its Monero Blockchain wallet RPC. Basically here is source code of it:

https://github.com/monero-project/monero/blob/50d48d611867ffcd41037e2ab4fec2526c08a7f5/contrib/epee/src/http_auth.cpp#L222

**Comment by davidsh** [comment]
>It's not .net core. Its Monero Blockchain wallet RPC. 

I'm confused because you say this is an ASP.NET Core application.  Is it an ASP.NET Core application but the app is making outbound HTTP requests to that C++ wallet RPC server?

**Comment by arkoc** [comment]
> > It's not .net core. Its Monero Blockchain wallet RPC.
> 
> I'm confused because you say this is an ASP.NET Core application. Is it an ASP.NET Core application but the app is making outbound HTTP requests to that C++ wallet RPC server?

Yes, sorry. Maybe I phrased sentence wrongly. Exactly, This is asp.net core app which is doing outbound Http requests to C++ wallet rpc (which is using digest authentication)

**Comment by davidsh** [comment]
@arkoc 

>Note: We use WinHttpHandler because new SocketHttpHandler was returning 401 in local machine too.

The SocketsHttpHandler digest bug might be #37729 which is fixed with PR #38225.  You'll need to use our daily builds or try out .NET Core 3.0 Preview 7 next month for that fix.

**Comment by davidsh** [comment]
We believe this problem is fixed in SocketsHttpHandler in .NET Core 3.0.

Please try out .NET Core 3.0:
https://devblogs.microsoft.com/dotnet/announcing-net-core-3-0-preview-7/

If this is still a problem, please open a new issue. Thanks!
